src/pages/LoginWithValidateCode2FAPage.js

```diff
@@ -58,24 +61,39 @@ class ValidateLogin2FANewWorkspacePage extends Component {
     }
 
     componentDidMount() {
-        // If the user has an active session already, they need to be redirected straight to the new workspace page
+        // If the user has an active session already, they need to be redirected straight to the relevant page
         if (this.props.session.authToken) {
             // In order to navigate to a modal, we first have to dismiss the current modal. But there is no current
-            // modal you say? I know, it confuses me too. Without dismissing the current modal, if they user cancels
-            // out of the new workspace modal, then they will be routed back to
-            // /v/<accountID>/<validateCode>/new-workspace and we don't want that. We want them to go back to `/` and
-            // by calling dismissModal(), the /v/... route is removed from history so the user will get taken to `/`
+            // modal you say? I know, it confuses me too. Without dismissing the current modal, if the user cancels out
+            // of the new workspace modal, then they will be routed back to
+            // /v/<accountID>/<validateCode>/workspace/123/card and we don't want that. We want them to go back to `/`
+            // and by calling dismissModal(), the /v/... route is removed from history so the user will get taken to `/`
             // if they cancel out of the new workspace modal.
             Navigation.dismissModal();
-
-            if (_.isEmpty(this.props.betas)) {
-                setRedirectToWorkspaceNewAfterSignIn(true);
-            } else {
-                create();
+            if (Permissions.canUseFreePlan(this.props.betas)) {
+                this.rerouteToRelevantPage();
             }
         }
     }
 
+    componentDidUpdate() {
+        // Betas can be loaded a little after a user is authenticated, so check again if the betas have been updated
+        if (this.props.session.authToken && Permissions.canUseFreePlan(this.props.betas)) {
+            this.rerouteToRelevantPage();
+        }
+    }
+
+    rerouteToRelevantPage() {
+        if (this.props.route.name === SCREENS.LOGIN_WITH_VALIDATE_CODE_2FA_WORKSPACE_CARD) {
```

Comment:
This method could use some more comments as it's pretty confusing what's going on.

Reply:
I also think a switch statement would be more readable

Reply:
Those are good points! Thank you for reviewing it as well @tgolen that will help me in next reviews 👍